config-linux.md

```diff
@@ -455,6 +455,35 @@ The following parameters can be specified to set up the controller:
    }
 ```
 
+### <a name="configLinuxRDMA" />RDMA
+
+**`rdma`** (object, OPTIONAL) represents the cgroup subsystem `rdma`.
+For more information, see the kernel cgroups documentation about [rdma][cgroup-v1-rdma].
+
+The following parameters can be specified to set up the controller:
+
+* **`hca_device`** *(string, OPTIONAL)* - specifies the device name whose resources limit to be configured
```

Comment:
I'd rather keep this REQUIRED, because there's no point in supporting `"rdma": [{}]`.  It also lets you drop the later line about “You must specify a valid `hca_device`…”.

Reply:
ok. so if I understand correctly, "rdma": [{}] can be treated as illegal config.
"rdma": [{"hca_device": "m1x5_1"}] is valid. And other two parameters are optional anyway.

So you are saying if end user passes hca_device only then we create the rdma controller?
Won't Docker or other engines create all empty cgroup controllers?

Reply:
> so if I understand correctly, `"rdma": [{}]` can be treated as illegal config

That's what you'd get if you REQUIRED `hca_device` (which I think is a good idea for consistency with [`rlimits[].type`][1] ).

> `"rdma": [{"hca_device": "m1x5_1"}]` is valid…

As you have it now.  But if you [requiring at least one of `hca_handles` or `hca_objects` to be set][2], then that would be invalid too (and we'd be consistent with the [current block I/O approach][3]).

> So you are saying if end user passes `hca_device` only then we create the rdma controller?  Won't Docker or other engines create all empty cgroup controllers as `"rdma": [{}]`?

If you have `rdma` unset or set to `[{}]` (or `[{"hca_device": "m1x5_1"}]`, etc.), you're not adding new limits.  The spec position on that [is][4]:

> Runtimes MAY attach the container process to additional cgroup controllers beyond those necessary to fulfill the `resources` settings.

which means that, for a config that does not adjust limits via `linux.resources.rdma`, the runtime can attach the container process to the rdma controller at `cgroupsPath` or not as it sees fit.  See #493, #834, and opencontainers/runc#861 for more background.

In case it helps clarify, I think this section of text should look like:

```
Each entry has the following structure:

* **`hca_device`** *(string, REQUIRED)* - specifies the device name whose resources limit to be configured
* **`hca_handles`** *(uint32, OPTIONAL)* - specifies the maximum number of hca_objects in the cgroup for a specified device
* **`hca_objects`** *(uint32, OPTIONAL)* - specifies the maximum number of hca_handles in the cgroup for a specified device

You MUST specify at least one of `hca_handles` or `hca_objects` in a given entry, and MAY specify both.
```

To close another no-op loophole we could follow [this][5] and require `rdma` to, when set, contain at least one entry.  That would be new wording though, since the spec doesn't currently do that for any OPTIONAL property.

[1]: https://github.com/opencontainers/runtime-spec/blame/v1.0.1/config.md#L167
[2]: https://github.com/opencontainers/runtime-spec/pull/942#discussion_r155593193
[3]: https://github.com/opencontainers/runtime-spec/pull/942#discussion_r155593193
[4]: https://github.com/opencontainers/runtime-spec/blame/v1.0.1/config-linux.md#L197
[5]: https://github.com/opencontainers/runtime-spec/blame/v1.0.1/config-linux.md#L543

Reply:
section of text that you described looks good to me. If we follow the text you described, it closes the no-op loophole too, isn't it for rdma controller?

Reply:
> If we follow the text you described, it closes the no-op loophole too, isn't it for rdma controller?

You could still have `"rdma": []` unless you REQUIRE the array to have at least one entry.  We don't close that loophole for any other OTIONAL properties at the moment (e.g. `"mounts": []` is ok), so we'd need maintainer input on whether it was worth setting a new precedent.

Another issue is the possibility of duplicate `hca_device` entries.  My preferred approach to tgat would be to use an object:

```
"rdma": {
  "m1x5_1": {
    "hca_handles": 3
  }
}
```

My attempt to do that for rlimits was rejected (#583), but I'm not clear on why.  It's possible that maintainers would support an object for this new property.

Reply:
I wouldn't worry about duplicate entries in case of rdma, because caller who is setting duplicate entries knows what is set. Last accessed entry would be applied, so that is fine.

I think keeping "rdma": [] in sync with "mounts":[] is ok.
This also allows to update the entry at later point without the need to new cgroup and moving all container processes to that new cgroup. Because in that process, some process might get left out without freeze-resume operation as mentioned in https://github.com/opencontainers/runc/issues/861.
so "rdma": [] seems better idea to me and I actually prefer that because it allows more flexibility in future for dynamically created RDMA devices which is on horizon currently and I am actively involved on kernel side of it, but not yet ready enough to discuss in this forum.

I don't have any strong preference or objection to schema proposed in this PR and below one.
Both schema are fine to me. Below one is little more intuitive where its evident that limits are of the specified device. I would +1 for below schema regardless of consideration for duplication.

```
"rdma": {
  "m1x5_1": {
    "hca_handles": 3
  }
}
```

Reply:
> I wouldn't worry about duplicate entries in case of rdma, because caller who is setting duplicate entries knows what is set. Last accessed entry would be applied, so that is fine.

We worry about them for `rlimits` (#607) and `namespaces` (#597), I don't see why this would be different.  One use-case would be if one person writes the template, and another uses [oci-runtime-tool][1] or `jq` to edit a particular entry.  That editing is easier if you're using a map, because accessing a particular entry is easier, you don't need to worry about finding the last matching entry, and you don't have to decide what to do about earlier matching entries.

> This also allows to update the entry at later point without the need to new cgroup and moving all container processes to that new cgroup.

If there are semantic differences between leaving `rdma` unset and setting it to `[]` (or, with the object approach, `{}`), I think we need to be very clear about what they are.  And we probably need to adjust the Go property to use a pointer to an array, and to not be `omitempty`.  I don't think that path is going to be particularly friendly :p.

[1]: https://github.com/opencontainers/runtime-tools

Reply:
There is no mention in runc spec on processing order of array or list of objects so finding last matching entry is possibly one way.
But I prefer less ambiguous schema as you have described, so I would go forward with it.

how about below summary.

1. If rdma is unset - no need to create rdma controller.
2. If "rdma": [] is provided, create rdma controller. It is ok to create empty controller.
Advantages: 
(a) No need to create new rdma cgroup when limits are updated
(b) As described at start of https://github.com/opencontainers/runc/issues/861, code is lot simpler which doesn't need to migrate existing processes to new cgroup via freeze-resume approach
3. "rdma":[ {} } ] accepts below zero or more entries.
```
  "mlx5_1": {
    "hca_handles": 3
    "hca_objects": 3
  },
  "mlx5_2": {
    "hca_handles": 3
  },
```
If this looks reasonable starting point, I will update the spec.

Reply:
> If this looks reasonable starting point...

(2) would be different from how the spec handles the current cgroups, at least to my reading.  And it doesn't avoid migration when the user leaves `rdma` unset at creation (your case (1)) but later sets it in an update (unless the runtime always joins all controllers).  And Go handling would need to distinguish between unset and an empty map.  But mitigating migration would be useful.  I'm now cautiously optimistic about this direction, although it would be good to have some maintainers weigh in.

Reply:
> (2) would be different from how the spec handles the current cgroups, at least to my reading. 

Below sample creates both the cgroups test1 and test2, which I think is correct behavior what I was expecting in sync with (2). or I misunderstood your comment?

```
package main

import (
        "cgroups"
        specs "github.com/opencontainers/runtime-spec/specs-go"
)

func main() {
        limits := new(specs.LinuxPids)

        _, _ = cgroups.New(cgroups.V1, cgroups.StaticPath("/test1"),
                           &specs.LinuxResources{
                           Pids : limits},)

        _, _ = cgroups.New(cgroups.V1, cgroups.StaticPath("/test2"),
                           &specs.LinuxResources{
                           Pids : nil},)
}
```
Regarding updating limits, cgroups go code does create new cgroup if it doesn't exist and apply the limits. I am not sure of process migration part using freeze-resume.

Reply:
> ```
> import (
>         "cgroups"
> ```

What is `cgroups`?  With your snippet in `test.go`:

```console
$ go run test.go 
test.go:4:2: cannot find package "cgroups" in any of:
        /usr/lib/go/src/cgroups (from $GOROOT)
        /home/wking/.local/lib/go/src/cgroups (from $GOPATH)
```

It looks like reading in the JSON would be ok, but I can't find a convenient approach to consistent serialization:

```
$ cat test.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

type SomeType struct {
	SomeMap map[string]string `json:"someMap"`
}

func main() {
	for _, data := range []string{
		"{}",
		"{\"someMap\": {}}",
		"{\"someMap\": {\"a\": \"b\"}}",
	} {
		someType := SomeType{}
		err := json.Unmarshal([]byte(data), &someType)
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		out, err := json.Marshal(someType)
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		fmt.Printf("%s -> %v (nil? %t, serialized: %s)\n", data, someType, someType.SomeMap == nil, string(out))
	}
}
$ go run test.go
{} -> {map[]} (nil? true, serialized: {"someMap":null})
{"someMap": {}} -> {map[]} (nil? false, serialized: {"someMap":{}})
{"someMap": {"a": "b"}} -> {map[a:b]} (nil? false, serialized: {"someMap":{"a":"b"}})
```

So you can use `nil` comparison to distinguish between “unset” (`== nil`) and “set to an empty object” (`!= nil && len(…) == 0`).  But we don't want the literal `null` in the serialized output (#662) for the originally-unset case.  And if we use:

```
SomeMap map[string]string `json:"someMap,omitempty"`
```

both the “unset” and “set to an empty object” cases serialize to “unset”.  So what Go type should we use to round-trip the distinction between “unset” and “set to an empty object”?

Reply:
>What is cgroups? With your snippet in test.go:
I am sorry for not putting github path there.
It is
"github.com/containerd/cgroups"

Reply:
> Below sample creates both the cgroups test1 and test2, which I think is correct behavior what I was expecting in sync with (2)…
>
> ```
>        limits := new(specs.LinuxPids)
>
>        _, _ = cgroups.New(cgroups.V1, cgroups.StaticPath("/test1"),
>                           &specs.LinuxResources{
>                           Pids : limits},)
> ```

`pids` is not a great example, because [`limits` is REQUIRED][1].  That means you cannot access [case (2)][2], which would be `"pids": {}` (meaning “create the cgroup (if necessary) and attach the container process to the cgroup, but don't adjust any limits”).

> ```
>         _, _ = cgroups.New(cgroups.V1, cgroups.StaticPath("/test2"),
>                            &specs.LinuxResources{
>                            Pids : nil},)
> ```

This is [case (1)][2], where runtimes [MAY][3] create the cgroup (if necessary) and attach the container process to the cgroup.  Apparently	containerd/cgroups decides to do that.  But runtimes don't have to.  And not creating the cgroup (and not attaching the container process to the non-existent cgroup) is the lazy-cgroups approach discussed in opencontainers/runc#861.  If containerd/cgroups decided to implement the lazy-cgroups approach on `nil`, their [`Create`][4] would instead look like:

```go
func (p *pidsController) Create(path string, resources *specs.LinuxResources) error {
  resources.Pids == nil {
    return nil  // case (1), opt for lazy creation and user did not set "pids"
  }
  if err := os.MkdirAll(p.Path(path), defaultDirPerm); err != nil {
    return err
  }
  if resources.Pids.Limit == nil {
    return nil  // case (2), user set "pids": {}, and we're pretending limit is OPTIONAL and an *int64
  }
  // case (3), user set "pids": {"limit": …}
  return ioutil.WriteFile(
    filepath.Join(p.Path(path), "pids.max"),
    []byte(strconv.FormatInt(resources.Pids.Limit, 10)),
    defaultFilePerm,
  )
}
```

[1]: https://github.com/opencontainers/runtime-spec/blame/v1.0.1/config-linux.md#L448
[2]: https://github.com/opencontainers/runtime-spec/pull/942#discussion_r155914884
[3]: https://github.com/opencontainers/runtime-spec/blame/v1.0.1/config-linux.md#L197
[4]: https://github.com/containerd/cgroups/blob/29da22c6171a4316169f9205ab6c49f59b5b852f/pids.go#L31-L43

Reply:
> This is case (1), where runtimes MAY create the cgroup (if necessary) and attach the container process to the cgroup. Apparently containerd/cgroups decides to do that. But runtimes don't have to. And not creating the cgroup (and not attaching the container process to the non-existent cgroup) is the lazy-cgroups approach discussed in opencontainers/runc#861. If containerd/cgroups decided to implement the lazy-cgroups approach on nil, their Create would instead look like:

I agree to what you described.
However current scheme (1) of runc-spec and its implementation sounds reasonble to me and fits the requirements of rdma too without #861.
When https://github.com/opencontainers/runc/issues/861 scheme is adopted, rdma could possibly follow that.

In order to make progress, I propose to move forward with currently defined scheme and implementation of (1), (2) and (3).
(2) and (3) are already matches generic runc-spec definition.

(1) is implemented by all cgroups (not just pids) in https://github.com/containerd/cgroups.
If (1) needs change in runc-spec definition or wording, it is going to be different PR which will affect other resources too.

Reply:
> If (1) needs change in runc-spec definition or wording, it is going to be different PR which will affect other resources too.

That makes sense to me.  I had been under the impression (based on [this comment][1]), that you intended to start requiring cgroup creation (if necessary) and joining in the `"rmda": {}` case, which is not a requirement that the current spec has for other controllers.  Currently the spec, as I read it, allows runtimes to decide whether they want to create/join or not for both the “unset” (e.g. `"resources": {}`) and “set empty” cases (e.g. `"resources": {"memory": {}}`) because of [this clause][2].  If you're ok using that same logic (for now) for `rmda`, then I'm back to preferring the wording I floated [here][3].

[1]: https://github.com/opencontainers/runtime-spec/pull/942#discussion_r155914884
[2]: https://github.com/opencontainers/runtime-spec/blame/v1.0.1/config-linux.md#L197
[3]: https://github.com/opencontainers/runtime-spec/pull/942#discussion_r155908710

Reply:
>That makes sense to me. I had been under the impression (based on this comment), that you intended to start requiring cgroup creation (if necessary) and joining in the "rmda": {} case, which is not a requirement that the current spec has for other controllers. Currently the spec, as I read it, allows runtimes to decide whether they want to create/join or not for both the “unset” (e.g. "resources": {}) and “set empty” cases (e.g. "resources": {"memory": {}}) because of this clause. If you're ok using that same logic (for now) for rmda, then I'm back to preferring the wording I floated here.

Based on the clause you mention https://github.com/opencontainers/runtime-spec/blame/v1.0.1/config-linux.md#L197 and based on implementation at https://github.com/containerd/cgroups, which follows the MAY part, -> "rdma": [{}] and "rdma": [{"hca_device": "m1x5_1"}] is also valid similar to priorities array of objects. For rdma instead of array of objects, as you suggested, we will have map as 
```
"rdma": {
  "m1x5_1": {
    "hca_handles": 3
  }
}
```
Are you ok with that?

Reply:
> For rdma instead of array of objects, as you suggested, we will have map as
>
> ```
> "rdma": {
>   "m1x5_1": {
>     "hca_handles": 3
>   }
> }
> ```
>
> Are you ok with that?

Ah, right, that would change the wording.  I'm in favor of the object/map approach, yes.

Reply:
ok. Great. So let me revise the spec to have treat them as optional and as map instead of array.